cocos/renderer/frame-graph/Resource.h

```diff
@@ -0,0 +1,170 @@
+#include "renderer/gfx-base/GFXTexture.h"
+
+namespace cc {
+namespace framegraph {
+
+template <typename DescriptorType>
+struct ResourceDescriptorHasher final {
+    inline uint32_t operator()(DescriptorType const &desc) const noexcept {
+        return 1;
+    }
+};
+
+template <typename DeviceResourceType, typename DescriptorType>
+struct DeviceResourceCreator final {
+    inline DeviceResourceType *operator()(DescriptorType const &desc) const noexcept {
+        return nullptr;
+    }
+};
+
+template <typename DeviceResourceType, typename DescriptorType,
+          typename DeviceResourceCreatorType = DeviceResourceCreator<DeviceResourceType, DescriptorType>,
+          typename DescriptorHasherType      = ResourceDescriptorHasher<DescriptorType>>
+class Resource final {
+public:
+    using Allocator        = ResourceAllocator<DeviceResourceType, DescriptorType, DeviceResourceCreatorType>;
+    using DescriptorHash   = typename Allocator::DescriptorHash;
+    using DeviceResource   = DeviceResourceType;
+    using Descriptor       = DescriptorType;
+    using DescriptorHasher = DescriptorHasherType;
+
+    Resource() noexcept = default;
+    explicit Resource(Descriptor const &desc) noexcept;
+    ~Resource()                = default;
+    Resource(Resource const &) = default;
+    Resource(Resource &&)      = default;
+    Resource &operator=(Resource const &) = default;
+    Resource &operator=(Resource &&) = default;
+
+    void                       createTransient() noexcept;
+    void                       createPersistent() noexcept;
+    void                       destroyTransient() noexcept;
+    void                       destroyPersistent() noexcept;
+    inline DeviceResourceType *get() const noexcept;
+    inline Descriptor const &  getDesc() const noexcept;
+
+private:
+    void computeHash() noexcept;
+
+    Descriptor          _desc;
+    DescriptorHash      _hash{0};
+    DeviceResourceType *_deviceObject{nullptr};
+};
+
+//////////////////////////////////////////////////////////////////////////
+
+template <typename DeviceResourceType, typename DescriptorType, typename DeviceResourceCreatorType, typename DescriptorHasherType>
+Resource<DeviceResourceType, DescriptorType, DeviceResourceCreatorType, DescriptorHasherType>::Resource(Descriptor const &desc) noexcept
+: _desc(desc) {
+}
+
+template <typename DeviceResourceType, typename DescriptorType, typename DeviceResourceCreatorType, typename DescriptorHasherType>
+void Resource<DeviceResourceType, DescriptorType, DeviceResourceCreatorType, DescriptorHasherType>::createTransient() noexcept {
+    computeHash();
+    _deviceObject = Allocator::getInstance().alloc(_desc, _hash);
+}
+
+template <typename DeviceResourceType, typename DescriptorType, typename DeviceResourceCreatorType, typename DescriptorHasherType>
+void Resource<DeviceResourceType, DescriptorType, DeviceResourceCreatorType, DescriptorHasherType>::createPersistent() noexcept {
+    computeHash();
+
```

Comment:
createTransient and createPersistent have duplicate hash calculations. In addition, the words Transient and Persistent seem like these two methods should be used as a choice between one or the other, but as I've seen in other engines, they should be used as a pre and a post method in the Resource creation process.

Reply:
I think the transient/persist concept here is completely different from what you mean there. 

Transient resources are created/reused in object pools, while persistent resources are owned by dedicated resource instances, and imported as external resources to the graph every frame. So the hash calculation function could be the same since indeed, one resource can only either be transient or persistent.

Could you elaborate on your side too?